File: mercury/extension.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: extension.m
% Main author: nyc
%
% Arithmetic in simple algebraic extension fields K(α),
% where α is a root of an irreducible polynomial over K.
%
% Elements of K(α) are represented as polynomials in α of degree
% less than the degree of the minimal polynomial.
%
%---------------------------------------------------------------------------%

:- module extension.
:- interface.

:- import_module poly.

%---------------------------------------------------------------------------%

    % A simple algebraic extension K(α)/K defined by the minimal
    % polynomial of α over K.
    %
:- type ext_field(K)
    --->    ext_field(
                gen_min_poly :: poly(K),
                ext_degree   :: int,
                ext_name     :: string
            ).

    % An element of an extension field, represented as a polynomial
    % in the generator of degree < ext_degree.
    %
:- type ext_elem(K)
    --->    ext_elem(
                elem_poly  :: poly(K),
                elem_field :: ext_field(K)
            ).

    % Construct an extension field from an irreducible polynomial.
    %
:- func mk_ext_field(poly(K), string) = ext_field(K) <= field(K).

    % Embed a base field element into the extension.
    %
:- func embed(ext_field(K), K) = ext_elem(K) <= ring(K).

    % The generator α of the extension.
    %
:- func generator(ext_field(K)) = ext_elem(K) <= ring(K).

    % Extension field arithmetic.
    %
:- func ext_add(ext_elem(K), ext_elem(K)) = ext_elem(K) <= field(K).
:- func ext_sub(ext_elem(K), ext_elem(K)) = ext_elem(K) <= field(K).
:- func ext_neg(ext_elem(K)) = ext_elem(K) <= ring(K).
:- func ext_mul(ext_elem(K), ext_elem(K)) = ext_elem(K) <= field(K).
:- func ext_inv(ext_elem(K)) = ext_elem(K) <= field(K).
:- func ext_div(ext_elem(K), ext_elem(K)) = ext_elem(K) <= field(K).
:- func ext_pow(ext_elem(K), int) = ext_elem(K) <= field(K).

    % Equality test for extension elements.
    %
:- pred ext_eq(ext_elem(K)::in, ext_elem(K)::in) is semidet <= ring(K).

    % Check if a field is the sentinel (placeholder for literals).
    %
:- pred is_sentinel_field(ext_field(K)::in) is semidet.

    % Extended Euclidean algorithm for polynomials.
    % Returns {G, S, T} such that G = S*A + T*B, with G monic.
    %
:- func ext_gcd(poly(K), poly(K)) = {poly(K), poly(K), poly(K)}
    <= field(K).

    % ring/field instances for ext_elem(K), enabling field tower nesting.
    %
:- instance ring(ext_elem(K)) <= field(K).
:- instance field(ext_elem(K)) <= field(K).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module list.
:- import_module require.

%---------------------------------------------------------------------------%

mk_ext_field(P, Name) = ext_field(monic(P), degree(monic(P)), Name).

embed(Field, C) = ext_elem(const_poly(C), Field).

generator(Field) = ext_elem(mono_x, Field).

%---------------------------------------------------------------------------%
% Reduction modulo minimal polynomial
%---------------------------------------------------------------------------%

:- func reduce(ext_field(K), poly(K)) = poly(K) <= field(K).

reduce(Field, P) = Result :-
    ( if is_sentinel_field(Field) then
        Result = P
    else
        div_mod(P, gen_min_poly(Field), _, Result)
    ).

%---------------------------------------------------------------------------%
% Arithmetic
%---------------------------------------------------------------------------%

ext_add(ext_elem(A, F), ext_elem(B, _)) =
    ext_elem(reduce(F, add(A, B)), F).

ext_sub(ext_elem(A, F), ext_elem(B, _)) =
    ext_elem(reduce(F, sub(A, B)), F).

ext_neg(ext_elem(A, F)) =
    ext_elem(neg(A), F).

ext_mul(ext_elem(A, F), ext_elem(B, _)) =
    ext_elem(reduce(F, mul(A, B)), F).

ext_inv(ext_elem(A, F)) = Result :-
    ( if coeffs(A) = [] then
        unexpected($pred, "division by zero in extension field")
    else
        {_, S, _} = ext_gcd(A, gen_min_poly(F)),
        Result = ext_elem(reduce(F, S), F)
    ).

ext_div(A, B) = ext_mul(A, ext_inv(B)).

ext_pow(E, N) = Result :-
    ( if N = 0 then
        Result = embed(elem_field(E), ring_one)
    else if int.'<'(N, 0) then
        Result = ext_pow(ext_inv(E), int.'-'(0, N))
    else if int.even(N) then
        Half = ext_pow(E, int.'//'(N, 2)),
        Result = ext_mul(Half, Half)
    else
        Result = ext_mul(E, ext_pow(E, int.'-'(N, 1)))
    ).

ext_eq(ext_elem(A, _), ext_elem(B, _)) :-
    coeffs(A) = coeffs(B).

%---------------------------------------------------------------------------%
% Sentinel field
%---------------------------------------------------------------------------%

:- func sentinel_field = ext_field(K) <= ring(K).

sentinel_field = ext_field(zero_poly, 0, "<literal>").

is_sentinel_field(F) :- ext_degree(F) = 0.

:- func pick_field(ext_field(K), ext_field(K)) = ext_field(K).

pick_field(F, G) =
    ( if is_sentinel_field(F) then G else F ).

%---------------------------------------------------------------------------%
% Extended GCD
%---------------------------------------------------------------------------%

ext_gcd(A, B) = Result :-
    ext_gcd_loop(A, B, const_poly(ring_one), zero_poly,
        zero_poly, const_poly(ring_one), Result).

:- pred ext_gcd_loop(poly(K)::in, poly(K)::in,
    poly(K)::in, poly(K)::in, poly(K)::in, poly(K)::in,
    {poly(K), poly(K), poly(K)}::out) is det <= field(K).

ext_gcd_loop(R0, R1, S0, S1, T0, T1, Result) :-
    ( if coeffs(R1) = [] then
        ( if lead_coeff(R0, LC) then
            InvLC = field_div(ring_one, LC),
            G = scale(InvLC, R0),
            S = scale(InvLC, S0),
            T = scale(InvLC, T0)
        else
            G = zero_poly,
            S = S0,
            T = T0
        ),
        Result = {G, S, T}
    else
        div_mod(R0, R1, Q, R),
        S2 = sub(S0, mul(Q, S1)),
        T2 = sub(T0, mul(Q, T1)),
        ext_gcd_loop(R1, R, S1, S2, T1, T2, Result)
    ).

%---------------------------------------------------------------------------%
% Typeclass instances
%---------------------------------------------------------------------------%

:- instance ring(ext_elem(K)) <= field(K) where [
    (ring_zero = ext_elem(zero_poly, sentinel_field)),
    (ring_one = ext_elem(const_poly(ring_one), sentinel_field)),
    (ring_add(A, B) = R :-
        ext_elem(PA, FA) = A,
        ext_elem(PB, FB) = B,
        F = pick_field(FA, FB),
        R = ext_elem(reduce(F, add(PA, PB)), F)),
    (ring_sub(A, B) = R :-
        ext_elem(PA, FA) = A,
        ext_elem(PB, FB) = B,
        F = pick_field(FA, FB),
        R = ext_elem(reduce(F, sub(PA, PB)), F)),
    (ring_mul(A, B) = R :-
        ext_elem(PA, FA) = A,
        ext_elem(PB, FB) = B,
        F = pick_field(FA, FB),
        R = ext_elem(reduce(F, mul(PA, PB)), F)),
    (ring_negate(ext_elem(P, F)) = ext_elem(neg(P), F)),
    (ring_from_int(N) =
        ext_elem(const_poly(ring_from_int(N)), sentinel_field)),
    (ring_is_zero(ext_elem(P, _)) :- coeffs(P) = [])
].

:- instance field(ext_elem(K)) <= field(K) where [
    (field_div(A, B) = ext_div(A, B))
].

%---------------------------------------------------------------------------%
:- end_module extension.
%---------------------------------------------------------------------------%

File: haskell/algebraic-polynomials/src/Math/Field/Extension.hs
-- |
-- Module      : Math.Field.Extension
-- Description : Arithmetic in simple algebraic extension fields K(alpha)
-- Stability   : experimental
--
-- Implements arithmetic in simple algebraic extension fields
-- \(K(\alpha) \cong K[x] / (m_\alpha(x))\), where \(\alpha\) is a root of
-- an irreducible polynomial \(m_\alpha(x) \in K[x]\).
--
-- Elements of \(K(\alpha)\) are represented as polynomials in \(\alpha\) of
-- degree less than \(\deg m_\alpha\). Arithmetic is performed modulo
-- \(m_\alpha\), with multiplicative inverse computed via the extended
-- Euclidean algorithm.
--
-- 'ExtElem' has 'Num' and 'Fractional' instances, enabling natural
-- arithmetic syntax and nesting for field towers:
-- @ExtElem (ExtElem Rational)@ represents \(\mathbb{Q}(\alpha)(\beta)\).
module Math.Field.Extension
  ( ExtField(..)
  , ExtElem(..)
  , mkExtField
  , embed
  , generator
  , extAdd
  , extSub
  , extNeg
  , extMul
  , extInv
  , extDiv
  , extPow
  , extEq
  , isSentinelField
  , extGcd
  ) where

import Math.Polynomial.Univariate

-- | A simple algebraic extension \(K(\alpha)/K\) defined by the minimal
-- polynomial of \(\alpha\) over \(K\).
data ExtField k = ExtField
  { genMinPoly :: !(Poly k)
    -- ^ Minimal polynomial of the generator (monic, irreducible over \(K\)).
  , extDegree  :: !Int
    -- ^ Degree of the extension, equal to \(\deg m_\alpha\).
  , extName    :: String
    -- ^ Display name for the generator (used in pretty-printing).
  } deriving (Show)

-- | An element of an extension field \(K(\alpha)\), represented as a
-- polynomial in \(\alpha\) of degree \(< \deg m_\alpha\).
data ExtElem k = ExtElem
  { elemPoly  :: !(Poly k)
    -- ^ Polynomial representation in the generator.
  , elemField :: !(ExtField k)
    -- ^ The extension field this element belongs to.
  } deriving (Show)

-- | Construct an extension field from an irreducible polynomial and a
-- name for the generator.
--
-- The polynomial is made monic. It should be irreducible over \(K\)
-- (this is not checked).
--
-- >>> let f = mkExtField (mkPoly [-2, 0, 1] :: Poly Rational) "sqrt2"
-- >>> extDegree f
-- 2
mkExtField :: (Eq k, Fractional k) => Poly k -> String -> ExtField k
mkExtField p name =
  let mp = monicPoly p
  in ExtField
    { genMinPoly = mp
    , extDegree  = degree mp
    , extName    = name
    }

-- | Embed a base field element into the extension as a constant polynomial.
--
-- >>> embed field (3 :: Rational)  -- the element 3 in Q(alpha)
embed :: (Eq k, Num k) => ExtField k -> k -> ExtElem k
embed field c = ExtElem (constPoly c) field

-- | The generator \(\alpha\) of the extension field, represented as the
-- polynomial \(x\) modulo \(m_\alpha\).
generator :: Num k => ExtField k -> ExtElem k
generator field = ExtElem monoX field

-- | Reduce a polynomial modulo the minimal polynomial of the extension.
--
-- For sentinel fields (used by literal constructors), no reduction is performed.
reduce :: (Eq k, Fractional k) => ExtField k -> Poly k -> Poly k
reduce field p
  | isSentinelField field = p  -- no reduction needed for literals
  | otherwise = snd $ divModPoly p (genMinPoly field)

-- | Addition in \(K(\alpha)\).
extAdd :: (Eq k, Fractional k) => ExtElem k -> ExtElem k -> ExtElem k
extAdd (ExtElem a f) (ExtElem b _) = ExtElem (reduce f $ addPoly a b) f

-- | Subtraction in \(K(\alpha)\).
extSub :: (Eq k, Fractional k) => ExtElem k -> ExtElem k -> ExtElem k
extSub (ExtElem a f) (ExtElem b _) = ExtElem (reduce f $ subPoly a b) f

-- | Negation in \(K(\alpha)\).
extNeg :: (Eq k, Num k) => ExtElem k -> ExtElem k
extNeg (ExtElem a f) = ExtElem (scalePoly (-1) a) f

-- | Multiplication in \(K(\alpha)\): polynomial multiplication followed
-- by reduction modulo \(m_\alpha\).
extMul :: (Eq k, Fractional k) => ExtElem k -> ExtElem k -> ExtElem k
extMul (ExtElem a f) (ExtElem b _) = ExtElem (reduce f $ mulPoly a b) f

-- | Multiplicative inverse in \(K(\alpha)\) via the extended Euclidean
-- algorithm.
--
-- Finds \(s\) such that \(s \cdot a \equiv 1 \pmod{m_\alpha}\)
-- by computing \(\gcd(a, m_\alpha) = 1 = s \cdot a + t \cdot m_\alpha\).
--
-- Throws an error for the zero element.
extInv :: (Eq k, Fractional k) => ExtElem k -> ExtElem k
extInv (ExtElem (Poly []) _) = error "extInv: division by zero"
extInv (ExtElem a f) =
  -- Find s such that s*a = 1 (mod minPoly)
  -- by extended GCD: gcd(a, minPoly) = 1 = s*a + t*minPoly
  let (_, s, _) = extGcd a (genMinPoly f)
  in ExtElem (reduce f s) f

-- | Division in \(K(\alpha)\): @extDiv a b = a * b^(-1)@.
extDiv :: (Eq k, Fractional k) => ExtElem k -> ExtElem k -> ExtElem k
extDiv a b = extMul a (extInv b)

-- | Exponentiation in \(K(\alpha)\) by an integer exponent.
--
-- Uses binary exponentiation for positive exponents.
-- Negative exponents compute the inverse first.
extPow :: (Eq k, Fractional k) => ExtElem k -> Int -> ExtElem k
extPow e 0 = embed (elemField e) 1
extPow e n
  | n < 0     = extPow (extInv e) (-n)
  | even n    = let half = extPow e (n `div` 2) in extMul half half
  | otherwise = extMul e (extPow e (n - 1))

-- | Equality test for extension field elements: compares their polynomial
-- representations.
extEq :: (Eq k, Num k) => ExtElem k -> ExtElem k -> Bool
extEq (ExtElem a _) (ExtElem b _) = a == b

-- | A sentinel field used by 'fromInteger' and 'fromRational' in the
-- 'Num' and 'Fractional' instances.
--
-- Binary operations pick the non-sentinel field from either operand,
-- allowing expressions like @3 * alpha@ where @3@ is a literal and
-- @alpha@ is a generator.
sentinelField :: ExtField k
sentinelField = ExtField (Poly []) 0 "<literal>"

-- | Check whether a field is the sentinel (placeholder for literals).
--
-- Sentinel fields have degree 0 and are used internally by 'fromInteger'
-- and 'fromRational'.
isSentinelField :: ExtField k -> Bool
isSentinelField f = extDegree f == 0

-- | Pick the "real" field from two operands. Prefers non-sentinel.
pickField :: ExtField k -> ExtField k -> ExtField k
pickField f g
  | isSentinelField f = g
  | otherwise         = f

-- | Equality of extension field elements.
instance (Eq k, Fractional k) => Eq (ExtElem k) where
  (==) = extEq

-- | Arithmetic in \(K(\alpha)\) via 'Num'.
--
-- 'fromInteger' creates a literal element with a sentinel field;
-- the actual field is resolved when the literal participates in
-- arithmetic with a proper extension element.
--
-- 'abs' and 'signum' are not meaningful and will throw errors.
instance (Eq k, Fractional k) => Num (ExtElem k) where
  ExtElem a fa + ExtElem b fb =
    let f = pickField fa fb
    in ExtElem (reduce f $ addPoly a b) f
  ExtElem a fa * ExtElem b fb =
    let f = pickField fa fb
    in ExtElem (reduce f $ mulPoly a b) f
  negate (ExtElem a f) = ExtElem (scalePoly (-1) a) f
  abs    = error "ExtElem: abs not meaningful for field elements"
  signum = error "ExtElem: signum not meaningful for field elements"
  fromInteger n = ExtElem (constPoly (fromInteger n)) sentinelField

-- | 'Fractional' instance enables division and 'fromRational' literals.
instance (Eq k, Fractional k) => Fractional (ExtElem k) where
  recip = extInv
  fromRational r = ExtElem (constPoly (fromRational r)) sentinelField

-- | Extended Euclidean algorithm for polynomials.
--
-- @extGcd a b@ returns @(g, s, t)@ such that \(g = s \cdot a + t \cdot b\)
-- and \(g = \gcd(a, b)\), with \(g\) made monic.
--
-- Used internally by 'extInv' to compute multiplicative inverses.
extGcd :: (Eq k, Fractional k) => Poly k -> Poly k -> (Poly k, Poly k, Poly k)
extGcd a b = go a b (constPoly 1) zeroPoly zeroPoly (constPoly 1)
  where
    go r0 r1 s0 s1 t0 t1
      | unPoly r1 == [] =
          let lc = case leadCoeff r0 of Just c -> c; Nothing -> 1
              s  = scalePoly (1/lc) s0
              t  = scalePoly (1/lc) t0
              g  = scalePoly (1/lc) r0
          in (g, s, t)
      | otherwise =
          let (q, r) = divModPoly r0 r1
              s2 = subPoly s0 (mulPoly q s1)
              t2 = subPoly t0 (mulPoly q t1)
          in go r1 r s1 s2 t1 t2
